File: cpp/leetcodes/Maximum_Product_Subarray.cpp
/*
 * Find the contiguous subarray within an array (containing at least one number) which has the largest product.

For example, given the array [2,3,-2,4],
the contiguous subarray [2,3] has the largest product = 6.
 */

#include <iostream>
#include <vector>
#include <bits/stl_algo.h>
#include <set>

using namespace std;

class Solution {
public:
    int maxProduct(vector<int> &nums) {
        if (nums.empty()) {
            return 0;
        }
        if (nums.size() == 1) {
            return nums[0];
        }
        int left = -1;
        int res = INT32_MIN;
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] == 0) {
                res = max(res, 0);
                if (left >= 0) {
                    res = max(res, maxProductInGroup(nums, left, i));
                    left = -1;
                }
            } else {
                if (left < 0) {
                    left = i;
                }
            }
        }
        if (left >= 0 && left < nums.size()) {
            res = max(res, maxProductInGroup(nums, left, nums.size()));
        }
        return res;
    }

    int maxProductInGroup(vector<int> &nums, int left, int right) {
        if (left == right - 1) {
            return nums[left];
        }
        int firstNegative = -1;
        int lastNegative = -1;
        int negativeCount = 0;
        for (int i = left; i < right; i++) {
            if (nums[i] < 0) {
                if (firstNegative < 0) {
                    firstNegative = i;
                    lastNegative = i;
                } else {
                    lastNegative = i;
                }
                negativeCount++;
            }
        }
        if ((negativeCount & 1) == 0) {
            int product = 1;
            for (int i = left; i < right; ++i) {
                product *= nums[i];
            }
            return product;
        } else {
            int p1 = 1;
            for (int i = left; i < lastNegative; ++i) {
                p1 *= nums[i];
            }
            int p2 = 1;
            for (int j = firstNegative + 1; j < right; ++j) {
                p2 *= nums[j];
            }
            return max(p1, p2);
        }
    }

};

void test(vector<int> &nums, int expect) {
    Solution s;
    int res = s.maxProduct(nums);
    if (res == expect) {
        cout << "Pass!" << endl;
    } else {
        cout << "Fail!, expect: " << expect << ", actual: " << res << endl;
    }
}

int main() {
    Solution s;
    auto v = vector<int>{2, 3, -2, 4};
    test(v, 6);
    v = vector<int>{2, 3, -2, -2, -2};
    test(v, 24);
    v = vector<int>{-1};
    test(v, -1);
    v = vector<int>{0};
    test(v, 0);
    v = vector<int>{-3, -4};
    test(v, 12);
    v = vector<int>{-3, 0, -3};
    test(v, 0);

}

File: leetcode/medium/Maximum_Product_Subarray.py
# -*- coding: utf-8 -*-

"""
created by huash06 at 2015-04-16 20:48

Find the contiguous subarray within an array (containing at least one number) which has the largest product.

For example, given the array [2,3,-2,4],
the contiguous subarray [2,3] has the largest product = 6.

"""

__author__ = 'huash06'

import sys
import os
import itertools
import collections
import functools
import bisect


class Solution:
    # @param nums, an integer[]
    # @return an integer
    def maxProduct(self, nums):
        groupNums = []
        group = []
        maxMul = -100000000

        # 把数字从0隔开，分成多个组，计算每个组的最大乘积
        for num in nums:
            if num == 0:
                if group:
                    groupNums.append(group)
                    group = []
                maxMul = 0
            else:
                group.append(num)
        groupNums.append(group)

        for group in groupNums:
            maxMul = max(maxMul, self.maxProductInGroup(group))

        return maxMul

    def maxProductInGroup(self, group):
        if not group:
            return 0
        if len(group) == 1:
            return group[0]
        # print(group)

        negativeCount = 0
        firstNegative = -1
        lastNegative = -1

        for index, num in enumerate(group):
            if num < 0:
                negativeCount += 1
                if firstNegative < 0:
                    firstNegative = index
                    lastNegative = index
                else:
                    lastNegative = index
        mulLam = lambda x, y: x*y

        # 如果有偶数个负号，直接把所以得数字做乘法
        if negativeCount & 1 == 0:
            return reduce(mulLam, group)
        # 否则舍弃第一个负号左边的数字或者最后一个负号右边的数字
        else:
            l = reduce(mulLam, group[firstNegative+1:]) if firstNegative+1 < len(group) else 0
            r = reduce(mulLam, group[:lastNegative]) if lastNegative > 0 else 0
            return max(l, r)


s = Solution()
print(s.maxProduct([-2]))
print(s.maxProduct([-2, 0, -1]))
print(s.maxProduct([2, 3, -2, 4]))
print(s.maxProduct([12]))
print(s.maxProduct([100, 12, 0, 2, 3, -5, 5, 6, -7, 8, 9, -6]))